src/bindings/c/ov/src/ov_c_api.cpp

```diff
@@ -613,47 +618,91 @@ bool ov_model_is_dynamic(const ov_model_t* model) {
 
 std::vector<char*> split(std::string s) {
     std::vector<char*> result;
-    std::regex delimiter("[,]");
-    std::sregex_token_iterator tokens(s.cbegin(), s.cend(), delimiter, -1);
-    std::sregex_token_iterator end;
-    for (; tokens != end; ++tokens) {
-        result.push_back(str_to_char_array(*tokens));
+
+    std::regex delimiter("[^{},]+");
+    auto words_begin = std::sregex_iterator(s.begin(), s.end(), delimiter);
+    auto words_end = std::sregex_iterator();
+    for (auto iter = words_begin; iter != words_end; ++iter) {
+        result.push_back(str_to_char_array(iter->str()));
     }
     return result;
 }
 
-ov_status_e ov_partial_shape_init(ov_partial_shape_t* partial_shape, const char* str) {
-    if (!partial_shape || !str) {
+ov_status_e ov_partial_shape_init(ov_partial_shape_t** partial_shape_obj, const char* str) {
+    *partial_shape_obj = nullptr;
+    if (!str) {
         return ov_status_e::GENERAL_ERROR;
     }
+    ov_partial_shape_t* partial_shape = nullptr;
     try {
         std::string s = str;
-        std::regex reg("[^,0-9.?][^\\-1]");
+        std::regex reg("[^,0-9.?}{}][^\\-1]");
         bool res = std::regex_search(s, reg);
         if (res) {
             return ov_status_e::PARAMETER_MISMATCH;
         }
-        std::vector<char*> result = split(s);
-        partial_shape->ranks = result.size();
-        if (partial_shape->ranks > MAX_DIMENSION) {
-            return ov_status_e::GENERAL_ERROR;
+
+        partial_shape = new ov_partial_shape_t;
+        if (s == "?") {  // dynamic rank
+            partial_shape->rank = ov::Dimension::dynamic();
+        } else {  // static rank
+            std::vector<char*> result = split(s);
+            size_t cnt = result.size();
+            if (cnt > MAX_DIMENSION) {
+                delete partial_shape;
+                return ov_status_e::GENERAL_ERROR;
+            }
+            partial_shape->rank = ov::Dimension(cnt);
+            for (auto& item : result) {
+                std::string dim = item;
+                if (dim == "?" || dim == "-1") {
+                    partial_shape->dims.emplace_back(ov::Dimension::dynamic());
+                } else {
+                    const std::string range_divider = "..";
+                    size_t range_index = dim.find(range_divider);
+                    if (range_index != std::string::npos) {
```

Comment:
When parsing a dimension with a range of values, could you use the same approach as the Python API uses?
https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/src/pyopenvino/core/common.cpp#L148
In particular I mean using the `ov::Dimension::value_type` for the string->number conversion to make sure we use the wide-enough type for the numbers which might arrive in this string. Additionally you can use `-1` for the max value if after parsing this end of a range you get an empty string. With those changes you can create the final dimension like this:
`Dimension::value_type min;`
`Dimension::value_type max;`
`// all of the parsing goes here`
`partial_shape->dims.emplace_back({min, max});`